include/pf-applications/grain_tracker/ellipsoid.h

```diff
@@ -0,0 +1,306 @@
+    min_radius() const
+    {
+      return *std::min_element(radii.cbegin(), radii.cend());
+    }
+
+    Number
+    max_radius() const
+    {
+      return *std::max_element(radii.cbegin(), radii.cend());
+    }
+
+    template <class Archive>
+    void
+    serialize(Archive &ar, const unsigned int /*version*/)
+    {
+      ar &A;
+      ar &b;
+      ar &alpha;
+      ar &radii;
+      ar &norm;
+      ar &gamma;
+      ar &center;
+    }
+  };
+
+  namespace internal
+  {
+    template <typename Number>
+    std::pair<Number, Number>
+    solve_quadratic(Number a, Number b, Number c)
+    {
+      const auto D = b * b - 4 * a * c;
+
+      return {(-b - std::sqrt(D)) / (2 * a), (-b + std::sqrt(D)) / (2 * a)};
+    }
+
+    template <int dim, typename Number>
+    Number
+    find_new_t(const Ellipsoid<dim, Number> &E,
+               const Tensor<1, dim, Number> &c1,
+               const Tensor<1, dim, Number> &c2)
+    {
+      const auto Ac1 = E.A * c1;
+      const auto Ac2 = E.A * c2;
+
+      const auto a = 0.5 * (Ac1 * c1 - 2 * Ac1 * c2 + Ac2 * c2);
+      const auto b = Ac2 * c1 - Ac1 * c1 + E.b * c2 - E.b * c1;
+      const auto c = 0.5 * Ac1 * c1 + E.b * c1 + E.alpha;
+
+      const auto sol = solve_quadratic(a, b, c);
+
+      auto in_range = [](Number v) { return v >= 0. && v <= 1.; };
+
+      AssertThrow(!in_range(sol.first) || !in_range(sol.second),
+                  ExcMessage("Both solutions " + std::to_string(sol.first) +
+                             " and " + std::to_string(sol.second) +
+                             " are in the admissible range"));
+      AssertThrow(in_range(sol.first) || in_range(sol.second),
+                  ExcMessage("None of the solutions " +
+                             std::to_string(sol.first) + " and " +
+                             std::to_string(sol.second) +
+                             " is in the admissible range"));
+
+      return (in_range(sol.first)) ? sol.first : sol.second;
+    }
+
+    template <int dim, typename Number>
+    Number
+    theta(const Tensor<1, dim, Number> &v, const Tensor<1, dim, Number> &w)
+    {
+      return std::acos(v * w / (v.norm() + w.norm()));
+    }
+
+    template <int dim, typename Number>
+    bool
+    crit(const Tensor<1, dim, Number> &v,
+         const Tensor<1, dim, Number> &w,
+         const Number                  eps)
+    {
+      return std::pow(v * w, 2) > std::pow(1 - eps * eps / 2., 2) *
+                                    v.norm_square() * w.norm_square();
+    }
+  } // namespace internal
+
+  template <int dim, typename Number>
+  std::tuple<Number, unsigned int, bool>
+  distance(const Ellipsoid<dim, Number> &E1,
+           const Ellipsoid<dim, Number> &E2,
+           const bool                    test_relative  = true,
+           const bool                    test_curvature = false,
+           const Number                  tol            = 1e-6,
+           unsigned int                  max_iter       = 100)
```

Comment:
What is your experience? How many iterations do we need?

Reply:
If the 2 ellipsoids are quite far from each other, then one needs like 50-100 iterations, as seen in this test https://github.com/hpsint/hpsint/blob/96c9882460ebf234416352f51339dc498ae01a10/tests/ellipses_distance_3d_02.output. If the ellipsods are realtively cloase, i.e. the distance is not larger than the largest radius of the two ellipsoids, then 6-8 iterations are sufficient (see https://github.com/hpsint/hpsint/pull/681/files#diff-14c3578bfcfbfd114fcba2dc637a8d59fdc52879e7ea983864cbd9f55c35397a). 
In practice, I will at first check the distance handling both ellipses as spheres taking the largest radius of each ellipsoids as a redius of the corresponding sphere. So if they do not overlap, then it even does not make sense to check the ellipsoidal distances, but if they do, then it means that the ellipsods are quite close, that should keep the number of iterations low. Moreover, the tolerance can be even slightly reduced to `1e-3` instead of `1e-6`, which further reduces iterations.

Yea, I will add the comments all over the code.

Reply:
It is also worth noting that a single iteration is quite cheap: building the coefficients of 2 quadratic equations, solving them and then updating 2 reference points.